src/cache.cpp

```diff
@@ -201,105 +218,75 @@ std::bitset<32> Cache::_CvtToAddr(const ulint block_set) {
-        }
+        a_tag(addr, block_set);
 
         break;
     case full_associative:
-        for (uint i = 31, j = 28; i > (31ul - _bit_tag); --i, --j) {
-            addr[i] = _cache[block_set][j];
-        }
+        a_tag(addr, block_set);
+
         break;
     case set_associative:
         for (ulint i = (_bit_block), j = 0; i < (_bit_block + _bit_set);
              ++i, ++j) {
             addr[i] = index[j];
         }
+        a_tag(addr, block_set);
 
-        for (uint i = 31, j = 28; i > (31ul - _bit_tag); --i, --j) {
-            addr[i] = _cache[block_set][j];
-        }
         break;
     }
 
     return addr;
 }
 
-std::bitset<32> Cache::_Evicted(const std::bitset<32> &addr) {
-    _cur_addr = addr;
-    _has_evicted = false;
-    bool space(false);
+void Cache::Ready(const std::bitset<32> &addr) {
+    _current_addr = addr;
+    _current_block = 0;
+    _current_set = 0;
+    _has_space = false;
 
-    if (_IsHit(_cur_addr)) {
-        _poten_victim |= std::bitset<32>(0xffffffff);
-    } else {
+    _has_hit = _IsHit(_current_addr);
+
+    if (!_has_hit) {
         switch (_cache_setting.associativity) {
         case direct_mapped:
-            _current_block = _GetCacheIndex(_cur_addr);
-            if (_cache[_current_block][30] &&
-                !_CheckIdent(_cache[_current_block], _cur_addr)) {
+            if (!_IfBlockAvailable()) {
                 _has_evicted = true;
-                _poten_victim = this->_CvtToAddr(_current_block);
             }
             break;
         case full_associative:
-            for (uint i = 0; i < _cache_setting.num_block; ++i) {
-                if (!_cache[i][30]) {
-                    space = true;
-                    _poten_victim |= std::bitset<32>(0xffffffff);
-                    break;
-                }
-            }
-            if (space) {
-                break;
-            }
-            _has_evicted = true;
-            if (_cache_setting.replacement_policy == RANDOM) {
-                _poten_victim = this->_CvtToAddr(GetBlockByRandom());
-            } else if (_cache_setting.replacement_policy == LRU) {
-                // Do sth.
-            }
-            break;
         case set_associative:
-            _current_set = _GetCacheIndex(_cur_addr);
-
-            for (ulint i = (_current_set * _cache_setting.cache_sets);
-                 i < ((_current_set + 1)) * _cache_setting.cache_sets; i++) {
-                if (!_cache[i][30]) {
-                    space = true;
-                    _poten_victim |= std::bitset<32>(0xffffffff);
-                    break;
+            if (!_IfBlockAvailable()) {
+                _has_evicted = true;
+                if (_cache_setting.replacement_policy == RANDOM) {
+                    int _ = GetBlockByRandom();
+                    _++;
+                } else if (_cache_setting.replacement_policy == LRU) {
+                    int _ = GetBlockByLRU();
+                    _++;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable '_' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)